### ISSUE TITLE
Lobby

### ISSUE THREAD

**Issue by alketii** [OpenDungeons/OpenDungeons#777]
An in-game lobby with chat could be very useful to play multiplayer.

**Comment by akien-mga** [comment]
Yeah mean basically adding a chat box in the current "seat configuration" screen?

**Comment by alketii** [comment]
I think before that, before hosting. I forgot to mention that lobby should be in masterserver (also requestion a masterserver :smile: )

**Comment by hwoarangmy** [comment]
yep. That means a need for a server ^^

**Comment by Bertram25** [comment]
@alketii I guess you mean a true lobby server, like in wesnoth, I suppose?

I'm not against the feature at all. In fact, it would be awesome! But I wondered, who could host this for us?
@Wout-van-Dongen Do you think it's possible for host this somehow (once implemented)?

**Comment by hwoarangmy** [comment]
@alketii Sorry for not being there yesterday but it was too late to connect when I was available. I will try to catch you on IRC.

@Bertram25 I talked with @alketii and he proposed to ask http://tuxfamily.org/ if they can host OD. I've had a look at their website but couldn't really find technical infos about what we can do or not (running executables ?). I've also seen that they host stuff for Mageia so @akien-mga probably already knows them.

Moreover, @Wout-van-Dongen also asked for a server version so he might also have something in mind.

**Comment by Bertram25** [comment]
Ok, ack. :) 

N.B.: It will be cool but I'll keep focusing on my current goals for now since it seems I'm slow enough those times. ;)

**Comment by alketii** [comment]
I'm willing to do the PHP/MySQL side.
What I need is:
0. To store the masterserver url in some ini
masterserver=http://opendungeons.tuxfamily.org/masterserver/
NOTE: this link doesn't exists , and ofc it should be used as variable in upcoming calls)
1. When someone clicks to host a game, an UUID should be generated and HTTP request should be made to 
   http://opendungeons.tuxfamily.org/masterserver/announce.php?uuid=UUID-HERE&ip=IP-HERE
2. Should make constant checks every 1 minute if server is still available
   http://opendungeons.tuxfamily.org/masterserver/update.php?uuid=UUID-HERE&status=0
3. When game starts
   http://opendungeons.tuxfamily.org/masterserver/update.php?uuid=UUID-HERE&status=1
4. Then during a game every 1 min another reuqest should be made
   http://opendungeons.tuxfamily.org/masterserver/update.php?uuid=UUID-HERE&status=1
5. And one when game ends
   http://opendungeons.tuxfamily.org/masterserver/update.php?uuid=UUID-HERE&status=2

ofc this is just a prototype and after this works, we can focus on adding live game stats, better security to prevent abuse etc. and this will show only in a web browser, but i will provide a CSV, json and XML (if you really need this last one), so it can be used in-game.

Legend of statuses 0 = Available, 1 = Game in progress, 2 = Game Over, 3 = Not responding
3 is checked in php/mysql

Thanks

**Comment by Bertram25** [comment]
@hwoarangmy @akien-mga Are we going with an web-based lobby server? (Just to make sure the scope and vision is ok here.)

If we're going toward that road:

@alketii I guess those external requests would end opening or ending dedicated server sessions, ready for players. Could you shortly describe what would happen behind each of those calls?
I.e: 
http://opendungeons.tuxfamily.org/masterserver/announce.php?uuid=MY_UUID
This would start a dedicated server with the given parameters:
--uuid=MY_UUID --pass=XXXX --port=YYYY --level=ZZZZZ

This would help us identify when the base logic has got flaws, so we can aim at a first stable process.

Thanks and best regards,

**Comment by alketii** [comment]
I just modified a bit the announce thing, added the most important part, the ip hehe, totally forgot abbout it.

@Bertram25 currently im making the php/mysql code, will publish very soon.

**Comment by alketii** [comment]
Also about the web thing, you can use it in game also, you can fetch the output of games in CSV format with getCSV.php and render it in game.

**Comment by alketii** [comment]
Here it is https://github.com/alketii/OpenDungeons-Masterserver
I also modified the "http requests doc"

**Comment by hwoarangmy** [comment]
From what I can guess from @alketii API, the server only is a repository for running games (OD executable has to be run in a "host" computer).
That's the simpliest we can have and that will not avoid hosts to run a modified cheated version of OD.
@alketii Do you know if it is possible to run an executable on tuxfamily servers ? That would be the best way to avoid cheats and take full advantages of the full server architecture...

**Comment by alketii** [comment]
Sorry for confusion, im not very familiar with these terms. What I did was a list when you can join games (so no need to exchange ip and will make finding other players quite easy)

Something like top part of this https://i.imgur.com/XFZGsso.jpg

**Comment by akien-mga** [comment]
> From what I can guess from @alketii API, the server only is a repository for running games (OD executable has to be run in a "host" computer).
> That's the simpliest we can have and that will not avoid hosts to run a modified cheated version of OD.

As a start though, it might be an interesting enough feature to let players find each other. If we want to prevent cheat, we might have an array of allowed maps hosted on the master server with md5/sha1/sha256/whichever-we-like-best sums to check if the map selected by the host is "official" or "custom" (I think we should allow custom maps, but warn about it in the lobby).

> @alketii Do you know if it is possible to run an executable on tuxfamily servers ? That would be the best way to avoid cheats and take full advantages of the full server architecture...

TuxFamily is already hosting some game servers as shown here: http://faq.tuxfamily.org/Games/En

**Comment by hwoarangmy** [comment]
> If we want to prevent cheat

As long as the host is on a player computer, there is no way to prevent cheat. It is so easy that I wouldn't even try to prevent it...

**Comment by alketii** [comment]
I think first is mostly important to find other players through masterserver, cheating could be prevented with headless server hosted/maintained by OD team.

**Comment by Bertram25** [comment]
> I think first is mostly important to find other players through masterserver, cheating could be prevented with headless server hosted/maintained by OD team.

+1 Let's make this work first and bring fun to players. Cheat can be covered afterwards. :)

**Comment by akien-mga** [comment]
So, I've discussed a bit with guys from TuxFamily (Reid from manaplus says hello btw @Bertram25 :p).
So basically they can provide the needed server facilities if we want to go with the mysql/php masterserver proposal, and they can also provide a Debian VM to host standalone server(s) directly.

For the VM, we'd need to have OD packaged for Debian. Given our currently development pace, I'm not sure an official inclusion in Debian main would be the way to go, as they are still lagging behind for cegui, etc., and they probably wouldn't update OD at each new release. But I think the TuxFamily guys would also be open to third-party deb files, and I guess @Bertram25 and I can manage that.

As I'm not really knowledgeable about servers, what we would need exactly from TuxFamily, etc., I think it would be best if @Bertram25 and/or @hwoarangmy could join the TuxFamily IRC channel to discuss things further with the guys there. It's not on Freenode though, it's on irc.tuxfamily.org/#tuxfamily.

**Comment by akien-mga** [comment]
And basically, I was told that as long as the game is free, they are really open to anything in terms of service that they can provide (at least that's how I understand it) :)

**Comment by Bertram25** [comment]
> Reid from manaplus says hello btw @Bertram25 :p

Eh eh, I miss her jokes, too.

> as they are still lagging behind for cegui, etc.

Muammar took care of that a few days ago ;)
https://ftp-master.debian.org/new/cegui-mk2_0.8.4-1.html

> But I think the TuxFamily guys would also be open to third-party deb files, and I guess @Bertram25 and I can manage that.

Totally, :D I think we could even adapt the control file and co from a close neighbor, such as 0ad or supertuxkart. :) 

> it would be best if @Bertram25 and/or @hwoarangmy could join the TuxFamily IRC channel to discuss things further with the guys there.

You might have noticed but I'm neck deep in RL stuff atm, so it will have to wait for next week as for me.
Anyway, before we start asking vm/servers for some services, I'd still like to know how you guys want to make the OD speak with the master server (which AFAIK is a different component from the actual lobby server, please note that.)

As a reminder and to me, the master server is small service maintaining a list of running games as @alketii described and implemented, and OD should then upload data of players hosting a WAN game to this server, while people looking for WAN games will then make a request to this server for the current list.

So, in order to be achieve all that, I'd suggest we:
- [x] Create a masterserver official repository and welcome @alketii 's code in it (Would you be ok to be officially part of the team @alketii ?) :)
- [ ] Decide what we'll use to send HTTP Post requests, Boost stream I suppose? http://stackoverflow.com/questions/11461962/sending-post-request
- [ ] Open an issue about implementing masterserver based WAN games support in OD and implement it.
  (Would it need a whole new mode, or would an option suffice in multiplayer?)
- Side-note: Once we have support to check for a list of available games online (a new screen would then be needed), I guess connecting to a general chatroom or a chat relay would be feasible and tada, you would have the start of your lobby. ;)

Once we've got those more or less done and being testing it locally, we'll be able to talk with our new (Tux)Family? ;)

In the meanwhile, we can still start to move our files there, anyway. But we can discuss that with them first.

WDYT guys?

**Comment by akien-mga** [comment]
> Muammar took care of that a few days ago ;)
> https://ftp-master.debian.org/new/cegui-mk2_0.8.4-1.html

Cool! Make sure to ping him to get https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=737058 closed.

About your proposal, it sounds really good :)

**Comment by Bertram25** [comment]
> Cool! Make sure to ping him to get https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=737058 closed.

You're right! I'll do that right now.

**Comment by alketii** [comment]
@Bertram25 yes I would like to join but I'm not sure what my contributions beside masterserver and testing would be since my knowledge on developing is limited.

**Comment by Bertram25** [comment]
@alketii Helping us maintaining the masterserver would be more than enough for a start. ;)
+ There might be other needs once we add a true lobby, etc...

@hwoarangmy @akien-mga @oyvindln @Danimal696 @Wout-van-Dongen 
Ok to create a new repo for the masterserver code and welcome @alketii in the developers team? :)

**Comment by akien-mga** [comment]
>  Ok to create a new repo for the masterserver code and welcome @alketii in the developers team? :)

Gladly :)

**Comment by hwoarangmy** [comment]
> Ok to create a new repo for the masterserver code and welcome @alketii in the developers team? :)

yep

@alketii Please note that we try to never merge our work (someone else has to review it and merge). That allows to be sure the dev branch is as stable as possible so please open a PR when you want to merge something.
Welcome in the team ^^

**Comment by Danimal696** [comment]
sure, fine by me

**Comment by alketii** [comment]
Thanks guys.

@hwoarangmy don't worry that I'm going to touch the development branch, just the masterserver.

**Comment by hwoarangmy** [comment]
> @hwoarangmy don't worry

I'm not worried ;-)

> I'm going to touch the development branch, just the masterserver

Well, if you want to do something in the source code or you feel like an artist and you want to mess with blender, you can. But even in the masterserver, you should do a PR that someone else will merge. Just as we will all do if we have something to change in the masterserver. The idea is that we should only merge code that as been reviewed. In OD source code as well as on the side projects (website, medias, masterserver, ...)

**Comment by Bertram25** [comment]
@alketii You've just been invited. Welcome. :)

@hwoarangmy  told everything I guess. :)

**Comment by Bertram25** [comment]
Ah, and the OpenDungeons-masterserver repo has been created.
https://github.com/OpenDungeons/OpenDungeons-masterserver

@alketii You should delete your repository and fork that one instead. The content will be the same + some readme fixes. :)

**Comment by alketii** [comment]
@hwoarangmy oh I misunderstood, thanks.

@Bertram25 deleted, thanks.

**Comment by Bertram25** [comment]
> The idea is that we should only merge code that as been reviewed. In OD source code as well as on the side projects (website, medias, masterserver, ...)

Ah, btw, I'd like to add this also permits to keep reviewers in the loop of everything happening.
But everyone can comment everyone's code. :)

**Comment by Bertram25** [comment]
Btw, meagaglest has got a masterserver in php. Might help: https://github.com/MegaGlest/megaglest-masterserver

**Comment by alketii** [comment]
I'm ready whenever you are :sunglasses:

**Comment by akien-mga** [comment]
Ok guys, back to this topic. It looks like @alketii is waiting on us, and we are waiting on him, so I guess we need to discuss how should this stuff be implemented.

We already have the masterserver repo; what is the next step?

Quoting @Bertram25 from a few comments above:
- [x] Create a masterserver official repository and welcome @alketii 's code in it (Would you be ok to be officially part of the team @alketii ?) :)
- [ ] Decide what we'll use to send HTTP Post requests, Boost stream I suppose? http://stackoverflow.com/questions/11461962/sending-post-request
- [ ] Open an issue about implementing masterserver based WAN games support in OD and implement it. (Would it need a whole new mode, or would an option suffice in multiplayer?)
- Side-note: Once we have support to check for a list of available games online (a new screen would then be needed), I guess connecting to a general chatroom or a chat relay would be feasible and tada, you would have the start of your lobby. ;)

For the first point, @alketii, is the masterserver code ready to send and process HTTP requests? Could we put a live instance on TuxFamily so that we can start trying to interact with it?

What about the second point? @hwoarangmy @Bertram25?

**Comment by alketii** [comment]
@akien-mga yes its ready to test, check the README for calls.

**Comment by alketii** [comment]
It's alive!!! http://od-bccb.rhcloud.com/

As you probably can see, we are not using TuxFamily :cry: . @akien-mga contacted them and they refused, he will probably explain.

**Comment by akien-mga** [comment]
Yeah basically I made a request on TF to open a MySQL database and a web area, with this description:

> We plan to use this combined with a MySQL database to host a PHP masterserver/lobby for our OpenDungeons game.
> 
> The masterserver code is under the GPLv3+ as the rest of the game, and located at: https://github.com/OpenDungeons/OpenDungeons-masterserver
> 
> The idea is that the OpenDungeons binary will send HTTP requests to the masterserver to populate an in-game lobby, letting players find each other. The actual game servers will still be hosted by users.

And it was rejected as they apparently don't host "project instances":

> The reason of refusal given by moderators is:
> As explained on http://faq.tuxfamily.org/Moderation/En (last bullet point of Moderation rules), TuxFamily hosts projects but not instances of those projects.

**Comment by hwoarangmy** [comment]
> The actual game servers will still be hosted by users.

? I thought the idea was to host the game on the server (at least, that was I was saying when talking about "hosting an executable)...

**Comment by akien-mga** [comment]
> > The actual game servers will still be hosted by users.
> 
> ? I thought the idea was to host the game on the server (at least, that was I was saying when talking about "hosting an executable)...

That's the very long term idea, but AFAICT there was more or less a consensus that we would start with the PHP masterserver that @alketii designed which would let users find each other more easily. The game hosting would be the same as it is currently, so basically this solution could be very easily implemented: we just need to process the data from the masterserver, return a list of servers waiting for players in a lobby UI, and let the player connect.

Hosting an executable at TF would mean packaging OpenDungeons for Debian Stable. That can take _years_. If we're friends with the TF guys, we could probably get them to accept a home made package to install on their Debian VM, but that's still some work. That's why I would prefer to have us implement the first solution (without necessarily polishing it too much, just something that works), and maybe work on external game server hosting in a second time.

**Comment by akien-mga** [comment]
The two main arguments in favour of the externally hosted game binary being:
- Avoid the need to open ports on the host for end users
- Avoid the risk of server-side cheating

The first argument is indeed quite desirable, as for the second I think it should not be a priority right now. When we have only a handful of players there is little concern about cheating :P

**Comment by alketii** [comment]
"hosting an executable" is known as Headless Server, which the first player who connect will control it (setup-wise) and costs money (depending on how many instances, it will be quite expensive). Headless server still has to be monitored by a masterserver (the current one which we have).

In the current one, when you click "Host a game" it will be made publicly available to the whole world, the players can find each other in the easiest way possible.

**Comment by hwoarangmy** [comment]
> The two main arguments in favour of the externally hosted game binary being:

I would add "Doing the server implementation once and for all instead of a temporary one"

Well, we have no php server and no VM to launch games. Maybe @Bertram25 will go for it but for my part, I won't do the server implementation as it seems we are not ready for it.

**Comment by akien-mga** [comment]
> Well, we have no php server and no VM to launch games.

Cf.

> It's alive!!! http://od-bccb.rhcloud.com/

_Edit:_ See also the csv-like version: http://od-bccb.rhcloud.com/get_csv.php

**Comment by akien-mga** [comment]
We also have https://github.com/OpenDungeons/OpenDungeons-masterserver which has been around for a while and powers the above php server.

**Comment by akien-mga** [comment]
> I would add "Doing the server implementation once and for all instead of a temporary one"

BTW how would you implement the lobby part in the hypothesis that we would have game servers hosted on TF or on a server that we would rent? Don't we need the php masterserver in any case? I don't see the above-discussed implementation as a temporary thing, only as a WIP as it might be brought to evolve _if_ we ever had VM hosted binaries. But most of the backend would stay the same I guess: tell a server that you are starting a game, give your IP and UUID, ask the server for the list of ongoing games.

**Comment by hwoarangmy** [comment]
> Don't we need the php masterserver in any case?

Yes we do and we would need to think about a way to launch the game somehow.

**Comment by Bertram25** [comment]
> Yes we do and we would need to think about a way to launch the game somehow.

I don't really know where the discussion is now, but please see the master server as a online list that can be checked to know what server is waiting for players or is running. (I guess you all understood, but I wanted to make it clear.)
Second point is that, whether we implement a true OD server or not, we can use the masterserver to make OD app in host mode to optionally tell the master server they are hosting a game and need players in any case.
then another app in join mode can check that same list and get the IP and port to connect to the server.
The only pb we have with this solution is that server behind routers will either need to see their router configured on the OD port to accept incoming packets and re-route them to the server, or we'll have to add techniques such as hole punching: https://en.wikipedia.org/wiki/Hole_punching_%28networking%29

In any case, the first part is relatively easy. :)
The second one is that hard, but somewhat more technical.

**Comment by alketii** [comment]
Its the first one, a simple list which can also be showed in game, and if a server has available slot, a player could join with one click.

**Comment by hwoarangmy** [comment]
I've been thinking a bit more about the lobby and I think we can have the same behaviour between the server hosted games and player hosted ones (excepted the source of the games).

@alketii I've been looking at your server API and we should adapt some stuff IMHO. If it is too late and you don't want to mess with that, please tell me and I will do it.
- The IP should be read by the server and not sent by the client (because we want the public IP, not the local one). I think we should be able to get it by using $_SERVER['REMOTE_ADDR']
- We are missing the port (and, this time, the client should set it).
- The UUID should be generated by the server rather than the client. The best being to mix a string witha timestamp and a random number.
- No need to update the game status during the game, only during lobby
- No need for more status than "available" and "over". For timeout, we only have to check for available games with a too long update.

**Comment by hwoarangmy** [comment]
> The only pb we have with this solution is that server behind routers will either need to see their router configured on the OD port to accept incoming packets

That will only be true for the server. And acceptable IMHO.

> we'll have to add techniques such as hole punching

Not needed IMHO as the goal is to have a server when we can

> Decide what we'll use to send HTTP Post requests, Boost stream I suppose?

I've had a quick look and I think SFML would be better. See:
http://www.sfml-dev.org/tutorials/2.3/network-http.php

**Comment by Bertram25** [comment]
Ok, fine by me.

**Comment by akien-mga** [comment]
> > we'll have to add techniques such as hole punching
> 
> Not needed IMHO as the goal is to have a server when we can

Just to clarify: I don't know what hole punching refers to, but regarding having our own server, I wouldn't count on it for now. We are still far away from having the possibility to host servers ourselves, and we should first focus on facilitating P2P IMO. If/when we have an own server, then we'll just have to reuse the P2P mechanisms with our server as one of the peers.

**Comment by hwoarangmy** [comment]
> I don't know what hole punching refers to

The problem with hosting a game (for a player) is that he needs to configure his router/firewall so that the other players can connect to his computer. On the orther side, for players that want to connect to your computer, it is pretty easy to connect as usually, high value ports are open.
Hole punching is a way to work around the problem for players hosting games.

Concerning our master server, I am having a look (should be ready soon) and IMHO, we can accept that players wanting to host a game have to setup their network to be able to host it.

**Comment by alketii** [comment]
Just a reminder: Current masterserver isn't secure in any way, It can be hacked, trolled, etc. by a simple web browser (or http request) :dancer: . I should fix it soon.

**Comment by hwoarangmy** [comment]
What do you have in mind ? Having to log to register a game ?
In any case, if you plan on doing something, could you please open an issue so that we know what you are on ?
In any case, if you want, I can give you the login/password for the master server
